lib/internal/fs/glob.js

```diff
@@ -0,0 +1,354 @@
+    }
+    this.#statsCache.set(path, val);
+    return val;
+  }
+  addToStatCache(path, val) {
+    this.#statsCache.set(path, val);
+  }
+  readdirSync(path) {
+    if (this.#readdirCache.has(path)) {
+      return this.#readdirCache.get(path);
+    }
+    let val;
+    try {
+      val = readdirSync(path, { __proto__: null, withFileTypes: true });
+    } catch {
+      val = [];
+    }
+    this.#readdirCache.set(path, val);
+    return val;
+  }
+  add(path, pattern) {
+    let cache;
+    if (this.#cache.has(path)) {
+      cache = this.#cache.get(path);
+    } else {
+      cache = new SafeSet();
+      this.#cache.set(path, cache);
+    }
+    const originalSize = cache.size;
+    pattern.indexes.forEach((index) => cache.add(pattern.cacheKey(index)));
+    return cache.size !== originalSize + pattern.indexes.size;
+  }
+  seen(path, pattern, index) {
+    return this.#cache.get(path)?.has(pattern.cacheKey(index));
+  }
+}
+
+class Pattern {
+  #pattern;
+  #globStrings;
+  indexes;
+  symlinks;
+  last;
+
+  constructor(pattern, globStrings, indexes, symlinks) {
+    this.#pattern = pattern;
+    this.#globStrings = globStrings;
+    this.indexes = indexes;
+    this.symlinks = symlinks;
+    this.last = pattern.length - 1;
+  }
+
+  isLast(isDirectory) {
+    return this.indexes.has(this.last) ||
+      (this.#pattern[this.last] === '' && isDirectory &&
+      this.indexes.has(this.last - 1) && this.#pattern[this.last - 1] === lazyMinimatch().GLOBSTAR);
+  }
+  isFirst() {
+    return this.indexes.has(0);
+  }
+  get hasSeenSymlinks() {
+    return ArrayPrototypeSome(ArrayFrom(this.indexes), (i) => !this.symlinks.has(i));
+  }
+  at(index) {
+    return this.#pattern[index];
+  }
+  child(indexes, symlinks = new SafeSet()) {
+    return new Pattern(this.#pattern, this.#globStrings, indexes, symlinks);
+  }
+  test(index, path) {
+    if (index > this.#pattern.length) {
+      return false;
+    }
+    const pattern = this.#pattern[index];
+    if (pattern === lazyMinimatch().GLOBSTAR) {
+      return true;
+    }
+    if (typeof pattern === 'string') {
+      return pattern === path;
+    }
+    if (typeof pattern?.test === 'function') {
+      return pattern.test(path);
+    }
+    return false;
+  }
+
+  cacheKey(index) {
+    let key = '';
+    for (let i = index; i < this.#globStrings.length; i++) {
+      key += this.#globStrings[i];
+      if (i !== this.#globStrings.length - 1) {
+        key += '/';
+      }
+    }
+    return key;
+  }
+}
+
+class Glob {
+  #root;
+  #exclude;
+  #cache = new Cache();
+  #results = [];
+  #queue = [];
+  constructor(patterns, options = kEmptyObject) {
+    validateObject(options, 'options');
+    const { exclude, cwd } = options;
+    if (exclude != null) {
+      validateFunction(exclude, 'options.exclude');
+    }
+    this.#root = cwd ?? '.';
+    this.#exclude = exclude;
+    this.matchers = ArrayPrototypeMap(patterns, (pattern) => new (lazyMinimatch().Minimatch)(pattern));
```

Comment:
This leaves a bit of optimization on the table.  Since we know that a glob walk will always be a canonical representation of the file (ie, you're matching against `a/b`, not `a/c/../b`), MM can prune out some `..` segments and optimize the handling of `**`.

```suggestion
    this.matchers = ArrayPrototypeMap(patterns, (pattern) => new (lazyMinimatch().Minimatch)(pattern, {
      nocase: process.platform === 'win32' || process.platform === 'darwin',
      nocaseMagicOnly: true,
      nocomment: true,
      nonegate: true,
      optimizationLevel: 2,
      platform: process.platform,
      windowsPathsNoEscape: true,
    }));
```

Reply:
Oh, btw, this isn't _just_ optimization. You definitely want to be doing case-insensitive matching on case-insensitive systems, or people will be confused that `echo test/*.JS` works on their cli, but passing `"test/*.JS"` to the test runner doesn't run anything. The `nocaseMagicOnly` prevents it from turning `a/b` into RegExps like `[/^a$/i, /^b$/i]` in nocase mode, since you're going to just let the file system do the case-insensitive matching for you.